etcd support broken

So I messed up `etcd` support the other day with PR #205. The last time etcd support worked was before my PR with git SHA `cc90c7b091275e606ad0ca7102a23fb2072f3f5e` of coreos/go-etcd (the deprecated lib). I'm working to fix but this is just so all are aware.

This is my current situation:

![what](https://user-images.githubusercontent.com/5950433/30883449-06539b72-a2c1-11e7-8694-62dd6f7921b9.png)


That failed:
```
pdxjohnny@pdxjohnny:kite$ KITE_TRANSPORT="WebSocket"  KONTROL_STORAGE="etcd" sudo -E make test
\033[0;32m==> Preparing test environment \033[0m
Using WebSocket transport
Cleaning /tmp/test_kite_home directory
Setting ulimit to 9000 for multiple client tests
\033[0;32m==> Using kontrol storage: 'etcd'\033[0m
Killing previous etcd instance
Installing etcd
test -d "_etcd" || git clone -b release-2.2 https://github.com/coreos/etcd _etcd
Creating test key
2017-09-26 13:49:03.907144 I | etcdmain: etcd Version: 2.2.5+git
2017-09-26 13:49:03.907183 I | etcdmain: Git SHA: 9f5189cd0
2017-09-26 13:49:03.907189 I | etcdmain: Go Version: go1.9
2017-09-26 13:49:03.907194 I | etcdmain: Go OS/Arch: linux/amd64
2017-09-26 13:49:03.907197 I | etcdmain: setting maximum number of CPUs to 4, total number of available CPUs is 4
2017-09-26 13:49:03.907201 W | etcdmain: no data-dir provided, using default data-dir ./default.etcd
2017-09-26 13:49:03.907450 I | etcdmain: listening for peers on http://localhost:2380
2017-09-26 13:49:03.907515 I | etcdmain: listening for peers on http://localhost:7001
2017-09-26 13:49:03.907597 I | etcdmain: listening for client requests on http://localhost:2379
2017-09-26 13:49:03.907668 I | etcdmain: listening for client requests on http://localhost:4001
2017-09-26 13:49:03.907896 I | etcdserver: name = default
2017-09-26 13:49:03.907909 I | etcdserver: data dir = default.etcd
2017-09-26 13:49:03.907916 I | etcdserver: member dir = default.etcd/member
2017-09-26 13:49:03.907921 I | etcdserver: heartbeat = 100ms
2017-09-26 13:49:03.907936 I | etcdserver: election = 1000ms
2017-09-26 13:49:03.907942 I | etcdserver: snapshot count = 10000
2017-09-26 13:49:03.907957 I | etcdserver: advertise client URLs = http://localhost:2379,http://localhost:4001
2017-09-26 13:49:03.907963 I | etcdserver: initial advertise peer URLs = http://localhost:2380,http://localhost:7001
2017-09-26 13:49:03.907975 I | etcdserver: initial cluster = default=http://localhost:2380,default=http://localhost:7001
2017-09-26 13:49:03.944313 I | etcdserver: starting member ce2a822cea30bfca in cluster 7e27652122e8b2ae
2017-09-26 13:49:03.944349 I | raft: ce2a822cea30bfca became follower at term 0
2017-09-26 13:49:03.944362 I | raft: newRaft ce2a822cea30bfca [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2017-09-26 13:49:03.944367 I | raft: ce2a822cea30bfca became follower at term 1
2017-09-26 13:49:03.944540 I | etcdserver: starting server... [version: 2.2.5+git, cluster version: to_be_decided]
2017-09-26 13:49:03.944758 E | etcdmain: failed to notify systemd for readiness: No socket
2017-09-26 13:49:03.944882 E | etcdmain: forgot to set Type=notify in systemd service file?
2017-09-26 13:49:03.945144 N | etcdserver: added local member ce2a822cea30bfca [http://localhost:2380 http://localhost:7001] to cluster 7e27652122e8b2ae
2017-09-26 13:49:04.344681 I | raft: ce2a822cea30bfca is starting a new election at term 1
2017-09-26 13:49:04.344725 I | raft: ce2a822cea30bfca became candidate at term 2
2017-09-26 13:49:04.344771 I | raft: ce2a822cea30bfca received vote from ce2a822cea30bfca at term 2
2017-09-26 13:49:04.344789 I | raft: ce2a822cea30bfca became leader at term 2
2017-09-26 13:49:04.344798 I | raft: raft.node: ce2a822cea30bfca elected leader ce2a822cea30bfca at term 2
2017-09-26 13:49:04.345107 I | etcdserver: published {Name:default ClientURLs:[http://localhost:2379 http://localhost:4001]} to cluster 7e27652122e8b2ae
2017-09-26 13:49:04.345156 I | etcdserver: setting up the initial cluster version to 2.2
2017-09-26 13:49:04.398714 N | etcdserver: set the initial cluster version to 2.2
\033[0;32m==> Downloading dependencies\033[0m
\033[0;32m==> Testing packages \033[0m
ok      github.com/koding/kite  20.079s
ok      github.com/koding/kite/config   1.015s
ok      github.com/koding/kite/dnode    1.008s
?       github.com/koding/kite/examples/exp2    [no test files]
?       github.com/koding/kite/examples/exp2-query      [no test files]
?       github.com/koding/kite/examples/math    [no test files]
?       github.com/koding/kite/examples/math/math-cli   [no test files]
?       github.com/koding/kite/examples/math/math-kite  [no test files]
?       github.com/koding/kite/examples/math-register   [no test files]
?       github.com/koding/kite/kitectl  [no test files]
?       github.com/koding/kite/kitectl/command  [no test files]
?       github.com/koding/kite/kitekey  [no test files]
?       github.com/koding/kite/kitetest [no test files]
exit status 1
FAIL    github.com/koding/kite/kontrol  0.060s
?       github.com/koding/kite/kontrol/kontrol  [no test files]
ok      github.com/koding/kite/kontrol/onceevery        3.005s
?       github.com/koding/kite/kontrol/protocol [no test files]
ok      github.com/koding/kite/protocol 1.005s
====> Starting WebSocket test
Starting kontrol
FAIL    github.com/koding/kite/reverseproxy     0.068s
?       github.com/koding/kite/reverseproxy/reverseproxy        [no test files]
ok      github.com/koding/kite/sockjsclient     1.212s
ok      github.com/koding/kite/systeminfo       1.048s
exit status 1
FAIL    github.com/koding/kite/test     0.052s
?       github.com/koding/kite/testkeys [no test files]
?       github.com/koding/kite/testutil [no test files]
?       github.com/koding/kite/testutil/writekey        [no test files]
Starting kontrol
FAIL    github.com/koding/kite/tunnelproxy      0.037s
?       github.com/koding/kite/tunnelproxy/tunnelproxy  [no test files]
?       github.com/koding/kite/utils    [no test files]
make: *** [Makefile:125: test] Error 1
```

It worked in travis though: https://travis-ci.org/koding/kite/jobs/279708934

--- a/kontrol/etcd.go
+++ b/kontrol/etcd.go
@@ -34,7 +34,7 @@ type Etcd struct {
 
 func NewEtcd(machines []string, log kite.Logger) *Etcd {
 	if machines == nil || len(machines) == 0 {
-		machines = []string{"//127.0.0.1:4001"}
+		machines = []string{"http://127.0.0.1:2379"}
 	}
 
 	cfg := etcd.Config{

--- a/kontrol/etcd.go
+++ b/kontrol/etcd.go
@@ -79,13 +79,13 @@ func (e *Etcd) Delete(k *protocol.Kite) error {
 	etcdKey := KitesPrefix + k.String()
 	etcdIDKey := KitesPrefix + "/" + k.ID
 
-	_, err := e.client.Delete(context.TODO(), etcdKey, &etcd.DeleteOptions{
+	_, e1 := e.client.Delete(context.TODO(), etcdKey, &etcd.DeleteOptions{
 		Recursive: true,
 	})
-	_, err = e.client.Delete(context.TODO(), etcdIDKey, &etcd.DeleteOptions{
+	_, e2 := e.client.Delete(context.TODO(), etcdIDKey, &etcd.DeleteOptions{
 		Recursive: true,
 	})
-	return err
+	return nonil(e1, e2)
 }
 
 func (e *Etcd) Clear() error {
@@ -95,29 +95,29 @@ func (e *Etcd) Clear() error {
 	return err
 }
 
-func (e *Etcd) Upsert(k *protocol.Kite, value *kontrolprotocol.RegisterValue) error {
-	return e.Add(k, value)
+func (e *Etcd) Upsert(k *protocol.Kite, v *kontrolprotocol.RegisterValue) error {
+	return e.Add(k, v)
 }
 
-func (e *Etcd) Add(k *protocol.Kite, value *kontrolprotocol.RegisterValue) error {
+func (e *Etcd) Add(k *protocol.Kite, v *kontrolprotocol.RegisterValue) error {
 	etcdKey := KitesPrefix + k.String()
 	etcdIDKey := KitesPrefix + "/" + k.ID
 
-	valueBytes, err := json.Marshal(value)
+	p, err := json.Marshal(v)
 	if err != nil {
 		return err
 	}
 
-	valueString := string(valueBytes)
+	value := string(p)
 
 	// Set the kite key.
 	// Example "/koding/production/os/0.0.1/sj/kontainer1.sj.koding.com/1234asdf..."
 	_, err = e.client.Set(context.TODO(),
 		etcdKey,
-		valueString,
+		value,
 		&etcd.SetOptions{
 			TTL:       KeyTTL,
-			PrevExist: etcd.PrevExist,
+			PrevExist: etcd.PrevIgnore,
 		},
 	)
 	if err != nil {
@@ -127,52 +127,46 @@ func (e *Etcd) Add(k *protocol.Kite, value *kontrolprotocol.RegisterValue) error
 	// Also store the the kite.Key Id for easy lookup
 	_, err = e.client.Set(context.TODO(),
 		etcdIDKey,
-		valueString,
+		value,
 		&etcd.SetOptions{
 			TTL:       KeyTTL,
-			PrevExist: etcd.PrevExist,
+			PrevExist: etcd.PrevIgnore,
 		},
 	)
-	if err != nil {
-		return err
-	}
 
-	return nil
+	return err
 }
 
-func (e *Etcd) Update(k *protocol.Kite, value *kontrolprotocol.RegisterValue) error {
+func (e *Etcd) Update(k *protocol.Kite, v *kontrolprotocol.RegisterValue) error {
 	etcdKey := KitesPrefix + k.String()
 	etcdIDKey := KitesPrefix + "/" + k.ID
 
-	valueBytes, err := json.Marshal(value)
+	p, err := json.Marshal(v)
 	if err != nil {
 		return err
 	}
 
-	valueString := string(valueBytes)
+	value := string(p)
 
 	// update the kite key.
 	// Example "/koding/production/os/0.0.1/sj/kontainer1.sj.koding.com/1234asdf..."
 	_, err = e.client.Set(context.TODO(),
 		etcdKey,
-		valueString,
+		value,
 		&etcd.SetOptions{
 			TTL:       KeyTTL,
 			PrevExist: etcd.PrevExist,
 		},
 	)
 	if err != nil {
-		err = e.Add(k, value)
-		if err != nil {
-			return err
-		}
-		return nil
+		// TODO(rjeczalik): Add only if err == KeyNotFound?
+		return e.Add(k, v)
 	}
 
 	// Also update the the kite.Key Id for easy lookup
 	_, err = e.client.Set(context.TODO(),
 		etcdIDKey,
-		valueString,
+		value,
 		&etcd.SetOptions{
 			TTL:       KeyTTL,
 			PrevExist: etcd.PrevExist,
@@ -191,11 +185,7 @@ func (e *Etcd) Update(k *protocol.Kite, value *kontrolprotocol.RegisterValue) er
 			PrevExist: etcd.PrevExist,
 		},
 	)
-	if err != nil {
-		return err
-	}
-
-	return nil
+	return err
 }
 
 func (e *Etcd) Get(query *protocol.KontrolQuery) (Kites, error) {
@@ -242,7 +232,6 @@ func (e *Etcd) Get(query *protocol.KontrolQuery) (Kites, error) {
 		KitesPrefix+"/"+etcdKey,
 		&etcd.GetOptions{
 			Recursive: true,
-			Sort:      false,
 		},
 	)
 	if err != nil {
@@ -289,7 +278,6 @@ func (e *Etcd) etcdKey(query *protocol.KontrolQuery) (string, error) {
 			KitesPrefix+"/"+query.ID,
 			&etcd.GetOptions{
 				Recursive: true,
-				Sort:      false,
 			},
 		)
 		if err != nil {

--- a/kontrol/kontrol.go
+++ b/kontrol/kontrol.go
@@ -588,3 +588,12 @@ func (k *Kontrol) generateToken(tok *token) (string, error) {
 
 	return signed, nil
 }
+
+func nonil(err ...error) error {
+	for _, e := range err {
+		if e != nil {
+			return e
+		}
+	}
+	return nil
+}

--- a/kontrol/kontrol/main.go
+++ b/kontrol/kontrol/main.go
@@ -81,8 +81,6 @@ func main() {
 	}
 
 	switch os.Getenv("KONTROL_STORAGE") {
-	case "etcd":
-		k.SetStorage(kontrol.NewEtcd(conf.Machines, k.Kite.Log))
 	case "postgres":
 		postgresConf := &kontrol.PostgresConfig{
 			Host:     conf.Postgres.Host,
@@ -95,6 +93,8 @@ func main() {
 		p := kontrol.NewPostgres(postgresConf, k.Kite.Log)
 		k.SetStorage(p)
 		k.SetKeyPairStorage(p)
+	case "etcd":
+		fallthrough
 	default:
 		k.SetStorage(kontrol.NewEtcd(conf.Machines, k.Kite.Log))
 	}

--- a/kontrol/kontrol/main.go
+++ b/kontrol/kontrol/main.go
@@ -58,7 +58,7 @@ func main() {
 
 	if conf.Initial {
 		initialKey(conf, publicKey, privateKey)
-		os.Exit(0)
+		return
 	}
 
 	kiteConf := config.MustGet()

